### PR TITLE
Create new PHPUnit assertions for the WebTestCase

### PR DESCRIPTION
**Pull request by Pierstoval**
| Q             | A
| ------------- | ---
| Branch?       | master
| Bug fix?      | no
| New feature?  | yes
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | yes
| License       | MIT
| Doc PR        | ~

I got inspirations from Laravel Dusk's assertions to bring new assertion capabilities to the `WebTestCase` in order to ease functional testing.

WDYT?

Edit: Here's the list of the added assertions, as a summary of this PR:

* `assertResponseIsSuccessful`
* `assertHttpCodeEquals`
* `assertResponseHasHeader`
* `assertResponseNotHasHeader`
* `assertResponseHeaderEquals`
* `assertResponseHeaderNotEquals`
* `assertResponseRedirects`
* `assertPageTitleEquals`
* `assertPageTitleContains`
* `assertClientHasCookie`
* `assertClientNotHasCookie`
* `assertClientCookieValueEquals`
* `assertClientRawCookieValueEquals`
* `assertResponseHasCookie`
* `assertResponseNotHasCookie`
* `assertResponseCookieValueEquals`
* `assertResponseCookieValueNotEquals`
* `assertSelectorExists`
* `assertSelectorNotExists`
* `assertSelectorContainsText`
* `assertSelectorNotContainsText`
* `assertInputValueEquals`
* `assertInputValueNotEquals`
* `assertRouteEquals`
* `assertRequestAttributeValueEquals`

### CODE REVIEW

File: src/Symfony/Bundle/FrameworkBundle/Test/WebTestCase.php [hunk 1]
```diff
@@ -41,6 +43,6 @@ protected static function createClient(array $options = [], array $server = [])
 
         $client->setServerParameters($server);
 
-        return $client;
+        return static::$client = $client;
```
Comment (stof) [inline src/Symfony/Bundle/FrameworkBundle/Test/WebTestCase.php]: if you store the client in the property, you must ensure you clean it on teardown so that we don't keep unwanted references to it.

File: src/Symfony/Bundle/FrameworkBundle/Test/WebTestAssertions.php [hunk 2]
```diff
@@ -0,0 +1,242 @@
...
+        );
+    }
+
+    public static function assertHttpCodeEquals(int $expectedCode): void
+    {
+        self::checkResponseAvailable();
+
+        $response = static::$client->getResponse();
+
+        Assert::assertSame(
```
[the remark this answers is not in the document]
Comment (Pierstoval) [reply src/Symfony/Bundle/FrameworkBundle/Test/WebTestAssertions.php]: Same vs equals is something that is specific to phpunit (and also to Twig), in PHP we make the difference by talking about strict or non-strict equality, so I think it's more convenient to stick to a more common approach of naming equality with "equals", WDYT?

File: src/Symfony/Bundle/FrameworkBundle/Test/WebTestCase.php (hunk 1 above)
Comment (Pierstoval) [reply src/Symfony/Bundle/FrameworkBundle/Test/WebTestCase.php]: Yep, done! Thanks :+1:

File: src/Symfony/Bundle/FrameworkBundle/Test/WebTestAssertions.php [hunk 3]
```diff
@@ -0,0 +1,292 @@
...
+            $expectedCode,
+            $code = $response->getStatusCode(),
+            "Response code [$expectedCode] does not match actual HTTP code [$code]."
+        );
+    }
+
+    public static function assertResponseRedirects(string $expectedLocation = null): void
```
Comment (mhujer) [inline src/Symfony/Bundle/FrameworkBundle/Test/WebTestAssertions.php]: @Pierstoval What about adding an expected redirect code (301 or 302)? I have it in my own `WebTestCase.php` - https://gist.github.com/mhujer/6754c021e0913e4b11c1788ac06ca768

File: src/Symfony/Bundle/FrameworkBundle/Test/WebTestAssertions.php [hunk 4]
```diff
@@ -0,0 +1,292 @@
...
+        );
+    }
+
+    public static function assertClientHasCookie(string $name, string $path = '/', string $domain = null): void
+    {
+        self::checkClientAvailable();
+
+        $cookie = static::$client->getCookieJar()->get($name, $path, $domain);
+
+        Assert::assertNotNull(
```
Comment (sstok) [inline src/Symfony/Bundle/FrameworkBundle/Test/WebTestAssertions.php]: Maybe return which cookies did exist? In case someone made a type.

File: src/Symfony/Bundle/FrameworkBundle/Test/WebTestAssertions.php [hunk 5]
```diff
@@ -0,0 +1,292 @@
...
+            $cookie,
+            "Did not find expected cookie [$name]."
+        );
+        Assert::assertSame(
+            $expectedValue,
+            $value = $cookie->getRawValue(),
+            "Cookie value [$expectedValue] does not match actual value [$value]."
```
Comment (sstok) [inline src/Symfony/Bundle/FrameworkBundle/Test/WebTestAssertions.php]: Mention the cookie name?

File: src/Symfony/Bundle/FrameworkBundle/Test/WebTestAssertions.php [hunk 6]
```diff
@@ -0,0 +1,292 @@
...
+        }
+    }
+
+    public static function assertPageTitleEquals(string $expectedTitle): void
+    {
+        self::checkCrawlerAvailable();
+
+        $titleNode = static::$client->getCrawler()->filter('title');
+
+        Assert::assertGreaterThan(0, $titleNode->count(), 'There is no <title> tag in the current page to look for.');
```
[the remark this answers is not in the document]
Comment (Pierstoval) [reply src/Symfony/Bundle/FrameworkBundle/Test/WebTestAssertions.php]: You're right, I'll fix this

File: src/Symfony/Bundle/FrameworkBundle/Test/WebTestAssertions.php (hunk 4 above)
Comment (Pierstoval) [reply src/Symfony/Bundle/FrameworkBundle/Test/WebTestAssertions.php]: Not sure, if the developer made a typo they will see it in the error message, since it's added in the assertion exception message

File: src/Symfony/Bundle/FrameworkBundle/Test/WebTestAssertions.php (hunk 5 above)
Comment (Pierstoval) [reply src/Symfony/Bundle/FrameworkBundle/Test/WebTestAssertions.php]: Yep, thanks for the catch!

File: src/Symfony/Bundle/FrameworkBundle/Test/WebTestAssertions.php [hunk 7]
```diff
@@ -0,0 +1,296 @@
...
+        $response = static::$client->getResponse();
+
+        Assert::assertSame(
+            $expectedCode,
+            $code = $response->getStatusCode(),
+            "Response code [$expectedCode] does not match actual HTTP code [$code]."
+        );
+    }
+
+    public static function assertResponseRedirects(int $expectedCode = null, string $expectedLocation = null): void
```
Comment (mhujer) [inline src/Symfony/Bundle/FrameworkBundle/Test/WebTestAssertions.php]: What about swapping the parameters? Having the `$expectedLocation` first will allow someone just to check that when the redirect code is not important.

File: src/Symfony/Bundle/FrameworkBundle/Test/WebTestAssertions.php [hunk 8]
```diff
@@ -0,0 +1,296 @@
...
+        self::checkResponseAvailable();
+
+        $response = static::$client->getResponse();
+
+        Assert::assertTrue(
+            $response->isRedirect(),
+            "Actual response code [{$response->getStatusCode()}] does not match any valid HTTP redirection code."
+        );
+
+        if ($expectedCode) {
```
Comment (mhujer) [inline src/Symfony/Bundle/FrameworkBundle/Test/WebTestAssertions.php]:
```suggestion
        if ($expectedCode !== null) {
```

File: src/Symfony/Bundle/FrameworkBundle/Test/WebTestAssertions.php [hunk 9]
```diff
@@ -0,0 +1,296 @@
...
+            static::assertHttpCodeEquals($expectedCode);
+        }
+
+        if ($expectedLocation) {
```
Comment (mhujer) [inline src/Symfony/Bundle/FrameworkBundle/Test/WebTestAssertions.php]:
```suggestion
        if ($expectedLocation !== null) {
```

Otherwise it will silently skip the check if someone accidentally puts `''` as an `$expectedLocation`.

File: src/Symfony/Bundle/FrameworkBundle/Test/WebTestAssertions.php [hunk 10]
```diff
@@ -0,0 +1,296 @@
...
+        }
+    }
+
+    public static function assertPageTitleEquals(string $expectedTitle): void
+    {
+        self::checkCrawlerAvailable();
+
+        $titleNode = static::$client->getCrawler()->filter('title');
+
+        Assert::assertSame(1, $titleNode->count(), 'There is no <title> tag in the current page to look for.');
```
Comment (mhujer) [inline src/Symfony/Bundle/FrameworkBundle/Test/WebTestAssertions.php]: The message may be misleading if there are two `<title>` tags. Maybe have to assertions? First check that there is at least one (with greaterThan and message `'There is no <title> tag in the current page to look for.'` and second `Assert::assertSame(1, $titleNode->count(), 'There must be exactly one <title> tag in the page.');`

Comment (mhujer) [reply src/Symfony/Bundle/FrameworkBundle/Test/WebTestAssertions.php]: or maybe an `assertCount` will work better?

File: src/Symfony/Bundle/FrameworkBundle/Test/WebTestAssertions.php (hunk 9 above)
Comment (Pierstoval) [reply src/Symfony/Bundle/FrameworkBundle/Test/WebTestAssertions.php]: A location that equals `""` is not something valid, so this should never happen

Comment (mhujer) [reply src/Symfony/Bundle/FrameworkBundle/Test/WebTestAssertions.php]: @Pierstoval It should never happen, but it can. But in case I accidentally put there `''` as expectedLocation (e.g. when i will have a list of URLs in the testcase and I'm iterating through them) I would prefer to know it immediately instead of letting it fail silently.

Comment (Pierstoval) [reply src/Symfony/Bundle/FrameworkBundle/Test/WebTestAssertions.php]: Well, I can change this so any non-null may fail in this case, seems fair

**Review by dunglas — APPROVED** [review]
Cool!

File: src/Symfony/Bundle/FrameworkBundle/Test/WebTestAssertions.php [hunk 11]
```diff
@@ -0,0 +1,296 @@
...
+ * Ideas borrowed from Laravel Dusk's assertions.
+ *
+ * @see https://laravel.com/docs/5.7/dusk#available-assertions
+ */
+trait WebTestAssertions
+{
+    /** @var Client|null */
+    protected static $client;
+
+    public static function assertResponseIsSuccessful(): void
```
Comment (localheinz) [inline src/Symfony/Bundle/FrameworkBundle/Test/WebTestAssertions.php]: Do these methods really need `public` visibility, or will `protected` suffice?

File: src/Symfony/Bundle/FrameworkBundle/Test/WebTestCase.php [hunk 12]
```diff
@@ -21,6 +21,15 @@
  */
 abstract class WebTestCase extends KernelTestCase
 {
+    use WebTestAssertions;
+
+    public function tearDown()
```
Comment (localheinz) [inline src/Symfony/Bundle/FrameworkBundle/Test/WebTestCase.php]: I believe `protected` visibility should suffice here. 

What do you think?

File: src/Symfony/Bundle/FrameworkBundle/Test/WebTestAssertions.php [hunk 13]
```diff
@@ -0,0 +1,296 @@
...
+
+    public static function assertResponseRedirects(string $expectedLocation = null, int $expectedCode = null): void
+    {
+        self::checkResponseAvailable();
+
+        $response = static::$client->getResponse();
+
+        Assert::assertTrue(
+            $response->isRedirect(),
+            "Actual response code [{$response->getStatusCode()}] does not match any valid HTTP redirection code."
```
Comment (localheinz) [inline src/Symfony/Bundle/FrameworkBundle/Test/WebTestAssertions.php]: What do you think about streamlining these messages?

See above in `assertResponseIsSuccessful`: 

```
"Response was expected to be successful, but actual HTTP code is {$response->getStatusCode()}."`
```

Perhaps here we use

```
"Response was expected to be a redirect, but actual HTTP code is {$response->getStatusCode()}."`
```

File: src/Symfony/Bundle/FrameworkBundle/Test/WebTestAssertions.php [hunk 14]
```diff
@@ -0,0 +1,296 @@
...
+        self::checkCrawlerAvailable();
+
+        $titleNode = static::$client->getCrawler()->filter('title');
+
+        Assert::assertSame(1, $count = $titleNode->count(), "There must be one <title> tag in the current page but there is actually $count.");
+
+        Assert::assertContains(
+            $expectedTitle,
+            \trim($title = $titleNode->text()),
+            "Expected title [$expectedTitle] does not equal actual title [$title]."
```
Comment (localheinz) [inline src/Symfony/Bundle/FrameworkBundle/Test/WebTestAssertions.php]: Should this be

```
"Expected title [$expectedTitle] does not contain actual title [$title]."
```

?

File: src/Symfony/Bundle/FrameworkBundle/Test/WebTestAssertions.php [hunk 15]
```diff
@@ -0,0 +1,296 @@
...
+
+    public static function assertSelectorNotContainsText(string $selector, string $text): void
+    {
+        self::checkCrawlerAvailable();
+
+        $nodes = static::$client->getCrawler()->filter($selector);
+
+        Assert::assertGreaterThan(0, $nodes->count(), "Selector [$selector] does not resolve to any node.");
+
+        Assert::assertContains($text, $nodes->text(), "Selector [$selector] does not contain text [$text].");
```
Comment (localheinz) [inline src/Symfony/Bundle/FrameworkBundle/Test/WebTestAssertions.php]: Should this be

```php
Assert::assertNotContains($text, $nodes->text(), "Selector [$selector] contains text [$text].");
```

?

File: src/Symfony/Bundle/FrameworkBundle/Test/WebTestAssertions.php [hunk 16]
```diff
@@ -0,0 +1,296 @@
...
+        self::checkCrawlerAvailable();
+
+        $inputNode = static::$client->getCrawler()->filter("input[name=\"$fieldName\"]");
+
+        Assert::assertGreaterThan(0, $inputNode->count(), "Input with name [$fieldName] not found on the page.");
+
+        Assert::assertNotEquals(
+            $expectedValue,
+            $value = $inputNode->getNode(0)->getAttribute('value'),
+            "Expected value [$value] for the [$fieldName] input does not equal the actual value [$value]."
```
Comment (localheinz) [inline src/Symfony/Bundle/FrameworkBundle/Test/WebTestAssertions.php]: Should this be

```
"Expected value [$value] for the [$fieldName] input equals the actual value [$value]."
```

File: src/Symfony/Bundle/FrameworkBundle/Test/WebTestAssertions.php (hunk 11 above)
Comment (Pierstoval) [reply src/Symfony/Bundle/FrameworkBundle/Test/WebTestAssertions.php]: I prefer to keep them public like the `Assert` class does, in case someone wants to test it or in case someone doesn't want to use the trait.

Comment (localheinz) [reply src/Symfony/Bundle/FrameworkBundle/Test/WebTestAssertions.php]: But how would you be able to test the trait _without_ using it?

File: src/Symfony/Bundle/FrameworkBundle/Test/WebTestCase.php (hunk 12 above)
Comment (Pierstoval) [reply src/Symfony/Bundle/FrameworkBundle/Test/WebTestCase.php]: Yep, didn't remember that the original method is `protected`, thanks 👍

File: src/Symfony/Bundle/FrameworkBundle/Test/WebTestAssertions.php (hunk 13 above)
Comment (Pierstoval) [reply src/Symfony/Bundle/FrameworkBundle/Test/WebTestAssertions.php]: Thanks for the suggestion, I'll add it 👍

File: src/Symfony/Bundle/FrameworkBundle/Test/WebTestAssertions.php (hunk 14 above)
Comment (Pierstoval) [reply src/Symfony/Bundle/FrameworkBundle/Test/WebTestAssertions.php]: Will fix, thanks 👍

File: src/Symfony/Bundle/FrameworkBundle/Test/WebTestAssertions.php (hunk 15 above)
Comment (Pierstoval) [reply src/Symfony/Bundle/FrameworkBundle/Test/WebTestAssertions.php]: Fixed, and I also reworded to `"Selector [$selector] was expected to not contain text [$text]."` instead, I think it is more explanatory 👍

File: src/Symfony/Bundle/FrameworkBundle/Test/WebTestAssertions.php (hunk 11 above)
Comment (Pierstoval) [reply src/Symfony/Bundle/FrameworkBundle/Test/WebTestAssertions.php]: Precisely by making its methods static and public 😄 
I just copied the behavior of the `Assert` class, that's all, and I think it's fine as-is

**Comment by XuruDragon** [comment]
Thank you ! so tired of always recoding the same bits of tests... This will greatly increase testing :)

**Comment by Pierstoval** [comment]
Thanks!

Friendly update to the reviewers: I added a slight summary of the assertions in the PR description so we can directly see what's added, for easier review.

Waiting for more reviews 👍

File: src/Symfony/Bundle/FrameworkBundle/Test/WebTestAssertions.php [hunk 17]
```diff
@@ -0,0 +1,413 @@
...
+
+    public static function assertResponseHeaderEquals(string $headerName, string $expectedValue): void
+    {
+        static::checkResponseAvailable();
+
+        $response = static::$client->getResponse();
+
+        Assert::assertSame(
+            $expectedValue,
+            $value = $response->headers->get($headerName),
```
Comment (ro0NL) [inline src/Symfony/Bundle/FrameworkBundle/Test/WebTestAssertions.php]: i think we should test all header values here, i.e. `get($headerName, null, true)`

File: src/Symfony/Bundle/FrameworkBundle/Test/WebTestAssertions.php [hunk 18]
```diff
@@ -0,0 +1,413 @@
...
+
+    public static function assertResponseHeaderNotEquals(string $headerName, string $expectedValue): void
+    {
+        static::checkResponseAvailable();
+
+        $response = static::$client->getResponse();
+
+        Assert::assertNotSame(
+            $expectedValue,
+            $value = $response->headers->get($headerName),
```
Comment (ro0NL) [inline src/Symfony/Bundle/FrameworkBundle/Test/WebTestAssertions.php]: and here as well

File: src/Symfony/Bundle/FrameworkBundle/Test/WebTestAssertions.php (hunk 17 above)
Comment (Pierstoval) [reply src/Symfony/Bundle/FrameworkBundle/Test/WebTestAssertions.php]: Thanks, solved!

File: src/Symfony/Bundle/FrameworkBundle/Test/WebTestAssertions.php [hunk 19]
```diff
@@ -0,0 +1,413 @@
...
+
+        $cookie = static::$client->getCookieJar()->get($name, $path, $domain);
+
+        Assert::assertNotNull(
+            $cookie,
+            "Did not find expected cookie [$name]."
+        );
+        Assert::assertSame(
+            $expectedValue,
+            $value = $cookie->getRawValue(),
```
Comment (renanbr) [inline src/Symfony/Bundle/FrameworkBundle/Test/WebTestAssertions.php]:
```suggestion
            $value = $cookie->getValue(),
```

Comment (Pierstoval) [reply src/Symfony/Bundle/FrameworkBundle/Test/WebTestAssertions.php]: Fixed 👍

File: src/Symfony/Bundle/FrameworkBundle/Test/WebTestAssertions.php [hunk 20]
```diff
@@ -0,0 +1,413 @@
...
+
+    public static function assertSelectorContainsText(string $selector, string $text): void
+    {
+        static::checkCrawlerAvailable();
+
+        $nodes = static::$client->getCrawler()->filter($selector);
+
+        Assert::assertGreaterThan(0, $nodes->count(), "Selector [$selector] does not resolve to any node.");
+
+        Assert::assertContains($text, $nodes->text(), "Selector [$selector] does not contain text [$text].");
```
Comment (renanbr) [inline src/Symfony/Bundle/FrameworkBundle/Test/WebTestAssertions.php]: Does it check against the first node of the list intentionally?
If not, is there a way to assert against all nodes?

same for `assertSelectorNotContainsText`

File: src/Symfony/Bundle/FrameworkBundle/Test/WebTestAssertions.php [hunk 21]
```diff
@@ -0,0 +1,413 @@
...
+    }
+
+    private static function checkResponseAvailable(): void
+    {
+        static::checkClientAvailable();
+
+        if (!static::$client->getResponse()) {
+            static::fail('A client must have an HTTP Response to make assertions.');
+        }
+    }
```
Comment (XuruDragon) [inline src/Symfony/Bundle/FrameworkBundle/Test/WebTestAssertions.php]: Maybe your `checkResponseAvailable()` could be more usefull by return the response than void?

As you do `static::$client->getResponse()` every time after `checkResponseAvailable()` call

File: src/Symfony/Bundle/FrameworkBundle/Test/WebTestAssertions.php [hunk 22]
```diff
@@ -0,0 +1,413 @@
...
+
+    private static function checkRequestAvailable(): void
+    {
+        static::checkClientAvailable();
+
+        if (!static::$client->getRequest()) {
+            static::fail('A client must have an HTTP Request to make assertions.');
+        }
+    }
```
Comment (XuruDragon) [inline src/Symfony/Bundle/FrameworkBundle/Test/WebTestAssertions.php]: Same than above but return the request

File: src/Symfony/Bundle/FrameworkBundle/Test/WebTestAssertions.php (hunk 21 above)
Comment (Pierstoval) [reply src/Symfony/Bundle/FrameworkBundle/Test/WebTestAssertions.php]: I removed the "check" methods and just use getters 👍

File: src/Symfony/Bundle/FrameworkBundle/Test/WebTestAssertions.php [hunk 23]
```diff
@@ -0,0 +1,382 @@
...
+
+    public static function assertResponseNotHasHeader(string $headerName): void
+    {
+        Assert::assertFalse(
+            static::getResponse()->headers->has($headerName),
+            "Header [$headerName] was not expected to be found in the Response."
+        );
+    }
+
+    public static function assertResponseHeaderEquals(string $headerName, string $expectedValue): void
```
Comment (ro0NL) [inline src/Symfony/Bundle/FrameworkBundle/Test/WebTestAssertions.php]: `array $expectedValue`

File: src/Symfony/Bundle/FrameworkBundle/Test/WebTestAssertions.php (hunk 20 above)
Comment (Pierstoval) [reply src/Symfony/Bundle/FrameworkBundle/Test/WebTestAssertions.php]: Yep, I think checking _all_ nodes might be a bit cumbersome, and it probably corresponds to non-common behavior

File: src/Symfony/Bundle/FrameworkBundle/Test/WebTestAssertions.php (hunk 23 above)
Comment (Pierstoval) [reply src/Symfony/Bundle/FrameworkBundle/Test/WebTestAssertions.php]: I'll go for `mixed` instead then, it can be different types

File: src/Symfony/Bundle/FrameworkBundle/Test/WebTestAssertions.php (hunk 11 above)
Comment (localheinz) [reply src/Symfony/Bundle/FrameworkBundle/Test/WebTestAssertions.php]: > Precisely by making its methods static and public 😄

Can you give a concrete example?

I don't think there is a need to unnecessarily use a visibility that is not needed.

File: src/Symfony/Bundle/FrameworkBundle/Test/WebTestAssertions.php (hunk 2 above)
[the remark this answers is not in the document]
Comment (localheinz) [reply src/Symfony/Bundle/FrameworkBundle/Test/WebTestAssertions.php]: I believe in intention revealing interfaces.

An invocation of this assertion will fail when `$expectedCode` is an `int` casted to a `string`, whereas - for example - an invokation of `assertEquals()` will pass.

**Review by localheinz — CHANGES_REQUESTED** [review]
Why are there no tests?

File: src/Symfony/Bundle/FrameworkBundle/Test/WebTestAssertions.php (hunk 11 above)
Comment (Pierstoval) [reply src/Symfony/Bundle/FrameworkBundle/Test/WebTestAssertions.php]: I have no special argument, I just think it's fine to be able to use it without using the trait

File: src/Symfony/Bundle/FrameworkBundle/Test/WebTestCase.php [hunk 24]
```diff
@@ -21,6 +21,15 @@
  */
 abstract class WebTestCase extends KernelTestCase
 {
+    use WebTestAssertions;
+
+    protected function tearDown()
+    {
```
Comment (stof) [inline src/Symfony/Bundle/FrameworkBundle/Test/WebTestCase.php]: you need to also call the parent method, otherwise the kernel is not shutdown properly

File: src/Symfony/Bundle/FrameworkBundle/Test/WebTestCase.php [hunk 25]
```diff
@@ -21,6 +21,15 @@
  */
 abstract class WebTestCase extends KernelTestCase
 {
+    use WebTestAssertions;
+
+    protected function tearDown()
```
Comment (stof) [inline src/Symfony/Bundle/FrameworkBundle/Test/WebTestCase.php]: this will cause issues with PHPUnit 8, while KernelTestCase was adapted to support it.

Comment (Pierstoval) [reply src/Symfony/Bundle/FrameworkBundle/Test/WebTestCase.php]: What would you recommend then?
Would it be ok to create another trait, like `RemoveClientOnShutdownTrait` or similar?

**Comment by Pierstoval** [comment]
@localheinz:
> Why are there no tests?

I guess because the `KernelTestCase` is not tested?

File: src/Symfony/Bundle/FrameworkBundle/Test/WebTestCase.php (hunk 25 above)
Comment (nicolas-grekas) [reply src/Symfony/Bundle/FrameworkBundle/Test/WebTestCase.php]: see #30474

Comment (Pierstoval) [reply src/Symfony/Bundle/FrameworkBundle/Test/WebTestCase.php]: Awesome! Thanks for the link, I'll do this 👍

File: src/Symfony/Bundle/FrameworkBundle/Test/WebTestCase.php (hunk 24 above)
Comment (Pierstoval) [reply src/Symfony/Bundle/FrameworkBundle/Test/WebTestCase.php]: Fixed in 86fdbb8

File: src/Symfony/Bundle/FrameworkBundle/Test/WebTestCase.php (hunk 25 above)
Comment (Pierstoval) [reply src/Symfony/Bundle/FrameworkBundle/Test/WebTestCase.php]: Fixed in 86fdbb8

**Comment by Pierstoval** [comment]
Friendly ping to @dunglas and @stof 

I fixed the compatibility with PHPUnit8 by using the same technique as https://github.com/symfony/symfony/pull/30474/files 

However, the aforementioned PR introduces an issue IMO: the `doSetUp()` and `doTearDown()` don't include the `: void` return type in the "PHPUnit <8 compatible" version of the class. Since Symfony 4 is compatible with PHP 7.1 only, I added the return type in all versions of the `TestCaseSetUpTearDownTrait` to be sure to not trigger the same issues than the one we have with PHPUnit8's BC break.

**Comment by Pierstoval** [comment]
The PR is ready, still waiting for reviews 😀

File: src/Symfony/Bundle/FrameworkBundle/Test/WebTestAssertions.php [hunk 26]
```diff
@@ -0,0 +1,382 @@
...
+            "Response was expected to be successful, but actual HTTP code is {$response->getStatusCode()}."
+        );
+    }
+
+    public static function assertHttpCodeEquals(int $expectedCode): void
+    {
+        Assert::assertSame(
+            $expectedCode,
+            $code = static::getResponse()->getStatusCode(),
+            "Response code [$expectedCode] does not match actual HTTP code [$code]."
```
Comment (fabpot) [inline src/Symfony/Bundle/FrameworkBundle/Test/WebTestAssertions.php]: We are using `""` instead of `[]` to quote things in Symfony.

File: src/Symfony/Bundle/FrameworkBundle/Test/WebTestAssertions.php [hunk 27]
```diff
@@ -0,0 +1,382 @@
...
+
+        return \reset($filteredCookies) ?: null;
+    }
+
+    private static function getClientForAssertion(): Client
+    {
+        if (!static::$client instanceof Client) {
+            static::fail(\sprintf(
+                'A client must be set to make assertions on it. Did you forget to call "%s::%s"?',
+                static::class, 'createClient'
```
Comment (fabpot) [inline src/Symfony/Bundle/FrameworkBundle/Test/WebTestAssertions.php]: createClient can be part of the string instead of an argument to `sprintf`

**Review by fabpot — COMMENTED** [review]
Looks interesting. I have not reviewed all assertions but I've made some cosmetic recommendations.

File: src/Symfony/Bundle/FrameworkBundle/Test/WebTestAssertions.php (hunk 26 above)
Comment (Pierstoval) [reply src/Symfony/Bundle/FrameworkBundle/Test/WebTestAssertions.php]: Thanks, fixed

File: src/Symfony/Bundle/FrameworkBundle/Test/WebTestAssertions.php (hunk 27 above)
Comment (Pierstoval) [reply src/Symfony/Bundle/FrameworkBundle/Test/WebTestAssertions.php]: Thanks for the catch, fixed

File: src/Symfony/Bundle/FrameworkBundle/Test/WebTestAssertions.php [hunk 28]
```diff
@@ -0,0 +1,382 @@
...
+    /** @var Client|null */
+    protected static $client;
+
+    public static function assertResponseIsSuccessful(): void
+    {
+        $response = static::getResponse();
+
+        Assert::assertTrue(
+            $response->isSuccessful(),
+            "Response was expected to be successful, but actual HTTP code is {$response->getStatusCode()}."
```
Comment (fabpot) [inline src/Symfony/Bundle/FrameworkBundle/Test/WebTestAssertions.php]: Idiomatic Symfony would be to use `sprintf()` here, like `sprintf('Response was expected to be successful, but actual HTTP code is "%s".', $response->getStatusCode())`. That would mean changing the other cases as well.

**Review by fabpot — CHANGES_REQUESTED** [review]

**Comment by Pierstoval** [comment]
Thanks @fabpot for the review, I fixed the cosmetic things.

fabbot.io's error seems like a false positive to me, since the phpdoc is necessary to sort of "warn" about the fact that the return type is void with PHPUnit 8.

I squashed everything, so the PR seems finished to me, unless anyone has another comment or idea I could work on 😄

File: src/Symfony/Component/Form/Test/TestCaseSetUpTearDownTrait.php [hunk 29]
```diff
@@ -45,17 +45,11 @@ protected function tearDown(): void
      */
     trait TestCaseSetUpTearDownTrait
     {
-        /**
-         * @return void
-         */
-        private function doSetUp()
+        private function doSetUp(): void
```
Comment (fabpot) [inline src/Symfony/Component/Form/Test/TestCaseSetUpTearDownTrait.php]: Why this change and the one on the Validator component? They don't seem related to your PR.

File: src/Symfony/Bundle/FrameworkBundle/Test/WebTestAssertions.php (hunk 28 above)
Comment (Pierstoval) [reply src/Symfony/Bundle/FrameworkBundle/Test/WebTestAssertions.php]: I will do this 👍

File: src/Symfony/Component/Form/Test/TestCaseSetUpTearDownTrait.php (hunk 29 above)
Comment (Pierstoval) [reply src/Symfony/Component/Form/Test/TestCaseSetUpTearDownTrait.php]: I squashed the commit with all the others, but this is intended to fix the fact that PHPUnit 8 needs the type-hint, and not PHPUnit 7.

As Symfony needs PHP 7.1, the void return type can be added with no harm, and since these traits are new as of the upcoming v4.3, IMO they can benefit from this return type right from the start.

I may separate the fix into another commit or into another PR if needed, but since I also need to add such trait in the new `WebTestAssertions`, I thought it would be also necessary to bring the fix to other traits at the same place. 

WYDT? Keep it as-is? New commit/PR?

Comment (nicolas-grekas) [reply src/Symfony/Component/Form/Test/TestCaseSetUpTearDownTrait.php]: For 4.2 and should be applied to all similar traits if there are more similar cases.

File: src/Symfony/Bundle/FrameworkBundle/Test/WebTestAssertions.php (hunk 28 above)
Comment (Pierstoval) [reply src/Symfony/Bundle/FrameworkBundle/Test/WebTestAssertions.php]: Done

File: src/Symfony/Component/Form/Test/TestCaseSetUpTearDownTrait.php (hunk 29 above)
Comment (Pierstoval) [reply src/Symfony/Component/Form/Test/TestCaseSetUpTearDownTrait.php]: So I should split the change and put it in another PR targeted on 4.2?

File: src/Symfony/Bundle/FrameworkBundle/Test/WebTestAssertions.php [hunk 30]
```diff
@@ -0,0 +1,383 @@
...
+            \sprintf('Response was expected to be successful, but actual HTTP code is %d.', $response->getStatusCode())
+        );
+    }
+
+    public static function assertHttpCodeEquals(int $expectedCode): void
+    {
+        Assert::assertSame(
+            $expectedCode,
+            $code = static::getResponse()->getStatusCode(),
+            \sprintf('Response code "%s" does not match actual HTTP code "%s".', $expectedCode, $code)
```
Comment (fabpot) [inline src/Symfony/Bundle/FrameworkBundle/Test/WebTestAssertions.php]: We don't need the `\\` before `sprintf`, we are only using it when it makes a performance difference, which is not the case here.

File: src/Symfony/Bundle/FrameworkBundle/Test/WebTestAssertions.php [hunk 31]
```diff
@@ -0,0 +1,383 @@
...
+
+    public static function assertPageTitleEquals(string $expectedTitle): void
+    {
+        $titleNode = static::getCrawler()->filter('title');
+
+        Assert::assertSame(1, $count = $titleNode->count(), \sprintf('There must be one <title> tag in the current page but there is actually %s.', $count));
+
+        Assert::assertEquals(
+            $expectedTitle,
+            \trim($title = $titleNode->text()),
```
Comment (fabpot) [inline src/Symfony/Bundle/FrameworkBundle/Test/WebTestAssertions.php]: no leading `\\`

File: src/Symfony/Bundle/FrameworkBundle/Test/WebTestAssertions.php [hunk 32]
```diff
@@ -0,0 +1,383 @@
...
+            $value = $request->attributes->get($key),
+            \sprintf('Expected request attribute "%s" value "%s" does not match actual value "%s".', $key, $expectedValue, $value)
+        );
+    }
+
+    protected static function getResponseCookieFromClient(string $name, string $path = '/', string $domain = null): ?Cookie
+    {
+        $cookies = static::getResponse()->headers->getCookies();
+
+        $filteredCookies = \array_filter($cookies, function (Cookie $cookie) use ($name, $path, $domain) {
```
Comment (fabpot) [inline src/Symfony/Bundle/FrameworkBundle/Test/WebTestAssertions.php]: Please remove all leading backslashes.

File: src/Symfony/Component/Form/Test/TestCaseSetUpTearDownTrait.php (hunk 29 above)
Comment (fabpot) [reply src/Symfony/Component/Form/Test/TestCaseSetUpTearDownTrait.php]: So, let's remove this change from here and let's create another PR for 4.2.

**Comment by fabpot** [comment]
@Pierstoval I think we also need some tests for these new assertions.

File: src/Symfony/Bundle/FrameworkBundle/Test/WebTestAssertions.php (hunk 30 above)
Comment (Pierstoval) [reply src/Symfony/Bundle/FrameworkBundle/Test/WebTestAssertions.php]: Done.

Is there any doc with the list of functions that are concerned by a performance boost with this backslash?

**Comment by Pierstoval** [comment]
@fabpot I'd be interested to know how I could test this, maybe I need a bit of coaching on this. I can create a single test and you could tell me if the direction is good or not 🙂

File: src/Symfony/Bundle/FrameworkBundle/Test/WebTestAssertions.php (hunk 30 above)
Comment (nicolas-grekas) [reply src/Symfony/Bundle/FrameworkBundle/Test/WebTestAssertions.php]: Here is the issue that references them: https://github.com/FriendsOfPHP/PHP-CS-Fixer/issues/3048
The process is the following: never add any, then let php-cs-fixer/fabbot add them where needed.

**Comment by fabpot** [comment]
@Pierstoval I've taken over this PR in #30813 where I've refactored the code and added tests.

**Comment by fabpot** [comment]
closing in favor of #30813
